### ISSUE TITLE
QB64: 32- and 64-bit BASIC Compiler and IDE

### ISSUE THREAD

**Issue by superloach** [skycocker/chromebrew#2569]
I've been attempting to compile QB64, with little success.

I modified the install script to use `bash setup_build.sh` instead of `./setup_build.sh`, and that fixed some initial errors. But now I keep getting this error, which is obviously related to #2567:
```
/usr/local/libexec/gcc/x86_64-cros-linux-gnu/4.9.4/cc1: error while loading shared libraries: libisl.so.15: cannot open shared object file: No such file or directory
```
Hopefully we can get that fixed, because I'd love to mess around with some of my old programs 😝

(p.s. - there's an SDL version of QB64 as well, in case there are graphical issues in future)

**Comment by superloach** [comment]
After moving the symlink to `libisl.so.19`, it complains about not finding `GL/glu.h`. I'll mention this on the QB64 forum, but should I open an issue on their GitHub?

**Comment by FellippeHeitor** [comment]
Hi and thanks for you interest in QB64. I just replied to your inquiry at our forum. The setup script attempts to detect the distro you're using to download the appropriate dependencies. You may be interested in trying to download them manually first. 
https://www.qb64.org/forum/index.php?topic=438.msg3140#msg3140

**Comment by FellippeHeitor** [comment]
Hope it works, please let us know.

**Comment by superloach** [comment]
It appears that `libglu1-mesa-dev` isn't in the repositories (or anything mentioning "glu", for that matter).

**Comment by superloach** [comment]
I have all the other requirements, though.
Doesn't appear that any other packages have provided it, either.
```
$ sudo find /usr/local/ -name *glu*
/usr/local/share/gtk-doc/html/gobject/glue.png
/usr/local/share/doc/groff-1.22.3/examples/chem/122/ch4z1_eqn_glutamic.chem
/usr/local/include/gssapi/mechglue.h
/usr/local/lib64/wine/fakedlls/glu32.dll
/usr/local/lib64/wine/libglu32.def
/usr/local/lib64/wine/glu32.dll.so
```
The only mention in this repository is a `--without-glu` flag in a PR for Wine.

**Comment by FellippeHeitor** [comment]
FreeGLUT is in the core of QB64. Looks like a deal breaker. We do have some legacy versions available at qb64.org that use SDL instead, as you mentioned above. Maybe one of those will cut it for you. Last version to use it was 0.954

**Comment by superloach** [comment]
I'll try the SDL version for now, but it's still possible we could get `libglu1-mesa-dev` in the repos and work from there.

**Comment by superloach** [comment]
The SDL (0.954) build fails with
```
In file included from libqbx.cpp:1:0:
common.cpp:33:23: fatal error: SDL_mixer.h: No such file or directory
 #include <SDL_mixer.h>
                       ^
compilation terminated.
g++: error: libqbx_lnx.o: No such file or directory
```

**Comment by FellippeHeitor** [comment]
😩 rats...

Well, good luck with the next option https://github.com/skycocker/chromebrew/issues/2569#issuecomment-414164597

**Comment by uberhacker** [comment]
Hey guys,
I was working on another project that required libglu / freeglut and just added 2 new PRs for those packages (#2570 and #2571), if it helps any.

**Comment by superloach** [comment]
Oh, that's awesome!!

**Comment by superloach** [comment]
```
$ crew -V
1.0.3
```
Also, I tried to `crew update`, and it gave these errors:
```
$ crew update
error: object file .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717 is empty
error: object file .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717 is empty
fatal: loose object 07bc22bff97ff3bffc2bf71f7d994808abe2c717 (stored in .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717) is corrupt
fatal: The remote end hung up unexpectedly
error: object file .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717 is empty
error: object file .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717 is empty
fatal: loose object 07bc22bff97ff3bffc2bf71f7d994808abe2c717 (stored in .git/objects/07/bc22bff97ff3bffc2bf71f7d994808abe2c717) is corrupt
Package lists, crew, and library updated.
Checking for package updates...
```
I'll just purge `/usr/local/*` and reinstall crew, hopefully that'll fix things.

**Comment by uberhacker** [comment]
Maybe try `sommelier -X qb64 --scale=1`?  You can also edit `/usr/local/sbin/sommelierd` and add the options you need for every login.

**Comment by superloach** [comment]
Sommelier doesn't seem to care about the `--scale` option.
```
$ sommelier -X qb64 --scale=1
Warning: Unknown scale factor: 1111
Warning: Unknown scale factor: 952
Warning: Unknown scale factor: 909
Warning: Unknown scale factor: 870
Warning: Unknown scale factor: 833
Warning: Unknown scale factor: 769
Warning: Unknown scale factor: 667
Child exited with status: 1
```

**Comment by superloach** [comment]
Forgot to mention, I get those errors with all GUI apps, and they work fine anyways. I think the issue may be that QB64 needs some of these files, but I haven't used a working build in a *while* so idk.
```
pwd;ls
/home/chronos/user/Downloads/qb64
cyberbit.ttf  LICENSE   qb64                             run_qb64.sh   source
internal      programs  'README for Linux and OS X.txt'  setup_lnx.sh  virtual_keyboard_layout_default.txt
```
Also `run_qb64.sh` doesn't work, because it uses `./`. Any idea why Chrome OS refuses to work when running things with that?

**Comment by uberhacker** [comment]
You might try `crew install qb64 -s > qb64.log 2>&1` and study the output of `qb64.log` to see if any messages stand out.

**Comment by superloach** [comment]
Uhh... `Package qb64 not found. :(`?

**Comment by uberhacker** [comment]
Oh, I thought you made a package already?  How did you "install"?

**Comment by FellippeHeitor** [comment]
QB64's binary must be able to find and write to the internal folder to work properly.

**Comment by superloach** [comment]
No, I just built it and copied the resulting binary to `/usr/local/bin`.
Ah yes, that's what I was thinking.

**Comment by superloach** [comment]
Sorry I'm not familiar with crew's packaging system, or really that much of Ruby for that matter. I'll definitely see if I can get a package together though, that might help things?

**Comment by uberhacker** [comment]
Packages can be as simple as just a few commands.  You can examine some of the other packages to get an idea but if you copy something that looks close to how you are trying install, it can act as a "template".  The wiki may help also.  See https://github.com/skycocker/chromebrew/wiki/Creating-a-package.

**Comment by superloach** [comment]
Alright, I've got a package made. Before I submit a PR, does this look right?
```ruby
require 'package'

class Qb64 < Package
  description 'QB64 is a 64-bit BASIC compiler and IDE.'
  homepage 'https://www.qb64.org/'
  version '1.2'
  source_url 'https://www.qb64.org/autobuilds/master/qb64_2018-02-28-12-16-57_c3ad985-master_lnx.tar.gz'
  source_sha256 '47c739847e3f0d517150eae28e459e4c2cc1f904c1ab11eb7d089810d56e044a'

  depends_on 'gcc7'
  depends_on 'libglu'
  depends_on 'freeglut'
  depends_on 'mesa'
  depends_on 'sommelier'
  depends_on 'alsa_lib'
  depends_on 'alsa_plugins'

  def self.build
    system "find . -name '*.sh' -exec sed -i \"s/\\r//g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\.\\/setup/bash setup/g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\.\\/qb64 \\&/exec qb64 \\&/g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\~\\/\.local\\/share\\/applicationexec qb64\\.desktop/\\/dev\\/null/g\" {} \\;"
    system "bash setup_lnx.sh"
  end

  def self.install
    system "chown", "chronos:chronos", "qb64"
    system "chmod", "+x", "qb64"
    system "mkdir", "-p", "#{CREW_DEST_PREFIX}/share/qb64"
    system "mkdir", "-p", "#{CREW_DEST_PREFIX}/bin"
    system "cp", "-rpa", ".", "#{CREW_DEST_PREFIX}/share/qb64/"
    system "echo '\#!/bin/bash' >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "echo \"cd #{CREW_PREFIX}/share/qb64/\" >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "echo \"sommelier -X #{CREW_PREFIX}/share/qb64/qb64\" >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "chmod +x #{CREW_DEST_PREFIX}/bin/qb64"
  end
end
```

**Comment by uberhacker** [comment]
Hey @superloach: You are definitely on the right track!  Some things I noticed:
- gcc7 and mesa are not needed as dependencies since mesa comes along for the ride in sommelier and gcc7 is already in "core"
- sommelier should be the last dependency listed to make sure the manual configuration steps are not missed
- You won't need the chown line since those permissions will already be inherited
- Instead of:

**Comment by superloach** [comment]
- Thanks for pointing that out about `gcc7` and `mesa`, I removed them.
- I put `sommelier` at the end, good call.
- Alright, I removed the `chown` line.
- It works without `sommelier -X`, but since some (like me) may choose to not run the daemon, and I already had to make a bash script to fix the internals dir, I thought I'd throw that in there for good measure.
- I'm unable to test this, but it *should* work on `i686` as well as `x86_64`. Should I go ahead and limit it to `x86_64` for now?
```ruby
require 'package'

class Qb64 < Package
  description 'QB64 is a modern extended BASIC+OpenGL language that retains QB4.5/QBasic compatibility and compiles native binaries for Windows, Linux and macOS.'
  homepage 'https://www.qb64.org/'
  version '1.2'
  source_url 'https://www.qb64.org/autobuilds/master/qb64_2018-02-28-12-16-57_c3ad985-master_lnx.tar.gz'
  source_sha256 '47c739847e3f0d517150eae28e459e4c2cc1f904c1ab11eb7d089810d56e044a'

  depends_on 'libglu'
  depends_on 'freeglut'
  depends_on 'alsa_lib'
  depends_on 'alsa_plugins'
  depends_on 'sommelier'

  def self.build
    system "find . -name '*.sh' -exec sed -i \"s/\\r//g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\.\\/setup/bash setup/g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\.\\/qb64 \\&/exec qb64 \\&/g\" {} \\;"
    system "find . -name '*.sh' -exec sed -i \"s/\~\\/\.local\\/share\\/applicationexec qb64\\.desktop/\\/dev\\/null/g\" {} \\;"
    system "bash setup_lnx.sh"
  end

  def self.install
    system "chmod", "+x", "qb64"
    system "mkdir", "-p", "#{CREW_DEST_PREFIX}/share/qb64"
    system "mkdir", "-p", "#{CREW_DEST_PREFIX}/bin"
    system "cp", "-rpa", ".", "#{CREW_DEST_PREFIX}/share/qb64/"
    system "echo '\#!/bin/bash' >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "echo \"cd #{CREW_PREFIX}/share/qb64/\" >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "echo \"sommelier -X #{CREW_PREFIX}/share/qb64/qb64\" >> #{CREW_DEST_PREFIX}/bin/qb64"
    system "chmod +x #{CREW_DEST_PREFIX}/bin/qb64"
  end
end
```
I also updated the description per [their homepage](https://www.qb64.org/).

**Comment by uberhacker** [comment]
i686 is 32-bit architecture so I'm not sure it will work but we can definitely try and find out.  The arm architectures are all 32-bit as well so it may only be limited to `x86_64`.

**Comment by superloach** [comment]
From what I can tell, QB64 still supports both 64- and 32-bit operating systems.
[FAQ: How do I upgrade the 32 bit Windows version of QB64 to 64 bit functionality?
](http://qb64.org/wiki/QB64_FAQ#Q:_How_do_I_upgrade_the_32_bit_Windows_version_of_QB64_to_64_bit_functionality.3F)
Hopefully this means it will work on ARM as well, we'll just have to see.
I'll go ahead and submit the PR with no architecture limits, and hopefully someone can test it out.

**Comment by uberhacker** [comment]
The project maintainer should leave out 64-bit in the description if it truly does work in 32-bit environments.  He could be sending the wrong message to more potential users of the app!  They may not even attempt to try it with the 64-bit mention.  But sure, go ahead and submit the PR.  I appreciate your efforts to package this app.

**Comment by superloach** [comment]
Well, when I opened this issue I forgot about 32-bit support, the actual description doesn't say that 😅

**Comment by superloach** [comment]
I'm also thinking of making a `qb64-dev` package, to use the latest development source. I threw together this little Bash one-liner just as an example of how I'd get the most recent one from the website.
```bash
$ echo "https://www.qb64.org/$(curl https://www.qb64.org/ 2>/dev/null|egrep -m 1 -o 'autobuilds\/development\/.*-development_lnx\.tar\.gz')"
https://www.qb64.org/autobuilds/development/qb64_2018-08-17-21-19-37_6cb81e1-development_lnx.tar.gz
```
Excuse my limited knowledge of Ruby, but what would be the easiest and safest way to implement something like this in the package definition? (Bonus: pull version numbers from resulting URL)

**Comment by uberhacker** [comment]
You can use the system command to execute any cli command you would normally run in the shell.  It would be tricky to package something like this since packages are supposed to have static versions instead of moving targets.  With that said, you can use backticks to execute a regular expression to grab the version and assign a variable: VERSION=``<regex pattern to extract from the url here>``

**Comment by superloach** [comment]
Thanks for the tips! I'll see what I can do.

**Comment by uberhacker** [comment]
Just tested on armv7l and received the following errors:
```
Building from source, this may take a while...
Rename all *.la files to *.la_tmp
Unable to detect distro, skipping package installation
Please be aware that for QB64 to compile, you will need the following installed:
  OpenGL developement libraries
  ALSA development libraries
  GNU C++ Compiler (g++)
Compiling and installing QB64...
rm: cannot remove './internal/temp/*': No such file or directory
Building library 'LibQB'
/tmp/cclw0Wye.s: Assembler messages:
/tmp/cclw0Wye.s:110584: Error: bad instruction `fldt [fp,#-36]'
/tmp/cclw0Wye.s:110584: Error: bad instruction `fistpll [fp,#-28]'
/tmp/cclw0Wye.s:110638: Error: bad instruction `fldt [fp,#-20]'
/tmp/cclw0Wye.s:110638: Error: bad instruction `fistpll [fp,#-12]'
/tmp/cclw0Wye.s:110670: Error: VFP single precision register expected -- `flds [fp,#-16]'
/tmp/cclw0Wye.s:110670: Error: bad instruction `fistpl [fp,#-8]'
/tmp/cclw0Wye.s:110701: Error: bad instruction `fldl [fp,#-20]'
/tmp/cclw0Wye.s:110701: Error: bad instruction `fistpl [fp,#-8]'
Building library 'FreeType'
Building library 'Core:FreeGLUT'
Building 'QB64'
g++: error: libqb/os/lnx/libqb_setup.o: No such file or directory
It appears that the qb64 executable file was not created, this is usually an indication of a compile failure (You probably saw lots of error messages pop up on the screen)
Usually these are due to missing packages needed for compilation. If you're not running a distro supported by this compiler, please note you will need to install the packages listed above.
If you need help, please feel free to post on the QB64 Forums detailing what happened and what distro you are using.
Also, please tell them the exact contents of this next line:
DISTRO: 

Thank you for using the QB64 installer.
Preconfiguring package...
chmod: cannot access 'qb64': No such file or directory
qb64 failed to install: `chmod +x qb64` exited with 1
```

**Comment by superloach** [comment]
Alright, should I limit it to just `x86_64` for now since `i686` and `armv7l` are both currently broken?

**Comment by uberhacker** [comment]
Either that or try and fix the issues on these architectures.  @FellippeHeitor?  Any suggestions?  Missing dependencies maybe?

**Comment by superloach** [comment]
The problem with `i686` appears to be an issue with graphics libraries or Sommelier.
The problem with `armv7l` appears to be instructions that aren't supported on ARM, which makes sense since this was intended for `x86`.

**Comment by superloach** [comment]
Oh, well that looks pretty simple! Hey @uberhacker, do you want me to go ahead and modify the package, or could you do it?

**Comment by uberhacker** [comment]
I will test and report back if successful.  Then you can make the change in the package.

**Comment by uberhacker** [comment]
Success!  Here is your patch:
```
diff --git a/packages/qb64.rb b/packages/qb64.rb
index 7cfdc6d6..cdf93caa 100644
--- a/packages/qb64.rb
+++ b/packages/qb64.rb
@@ -18,6 +18,7 @@ class Qb64 < Package
     system "find . -name '*.sh' -exec sed -i \"s/\.\\/setup/bash setup/g\" {} \\;"
     system "find . -name '*.sh' -exec sed -i \"s/\.\\/qb64 \\&/exec qb64 \\&/g\" {} \\;"
     system "find . -name '*.sh' -exec sed -i \"s/\~\\/\.local\\/share\\/applicationexec qb64\\.desktop/\\/dev\\/null/g\" {} \\;"
+    system "sed -i '1i#define QB64_NOT_X86' internal/c/common.h"
     system "bash setup_lnx.sh"
   end
```

**Comment by superloach** [comment]
Great, thanks a lot!

**Comment by uberhacker** [comment]
Actually, I thought of a more elegant way to do all this patching.  New patch coming soon...

**Comment by superloach** [comment]
Haha, alright then.

**Comment by uberhacker** [comment]
Try this instead:
```
diff --git a/packages/qb64.rb b/packages/qb64.rb
index 7cfdc6d6..d60628bf 100644
--- a/packages/qb64.rb
+++ b/packages/qb64.rb
@@ -13,11 +13,18 @@ class Qb64 < Package
   depends_on 'alsa_plugins'
   depends_on 'sommelier'

-  def self.build
+  def self.patch
     system "find . -name '*.sh' -exec sed -i \"s/\\r//g\" {} \\;"
     system "find . -name '*.sh' -exec sed -i \"s/\.\\/setup/bash setup/g\" {} \\;"
     system "find . -name '*.sh' -exec sed -i \"s/\.\\/qb64 \\&/exec qb64 \\&/g\" {} \\;"
     system "find . -name '*.sh' -exec sed -i \"s/\~\\/\.local\\/share\\/applicationexec qb64\\.desktop/\\/dev\\/null/g\" {} \\;"
+    case ARCH
+    when 'armv7l', 'aarch64'
+      system "sed -i '1i#define QB64_NOT_X86' internal/c/common.h"
+    end
+  end
+
+  def self.build
     system "bash setup_lnx.sh"
   end
```

**Comment by superloach** [comment]
Yeah, that's way more elegant! Sorry I didn't know `self.patch` was a thing, it wasn't used in the example I built off. 😛

**Comment by uberhacker** [comment]
Now if we can solve the mysterious black screen of death, then i686 will be working also.

**Comment by superloach** [comment]
Hmm, this is interesting: [Re: How do you make the 64 bit Version of QB64?](https://www.qb64.org/forum/index.php?topic=4.msg773#msg773)
This makes it sound like Unix only supports 64-bit now, but I'm not sure what version that was.

**Comment by FellippeHeitor** [comment]
That's an old FAQ; QB64 is 32bit by default on Windows; 64bit by default on macOS; 32/64bit on Linux depending on the Distro it's being compiled on, since it'll use the gcc version that's installed.

**Comment by superloach** [comment]
Thanks for clearing that up, I couldn't find a straight answer anywhere!

**Comment by FellippeHeitor** [comment]
I for instance have it installed on two separate VMs for occasional testing, both running Debian, one 32 one 64 bit.

**Comment by uberhacker** [comment]
@FellippeHeitor: The good news is we now have it working in ChromeOS on both x86_64 and arm architecture which probably covers over 95% of all Chromebook owners currently.

**Comment by FellippeHeitor** [comment]
That's amazing. We can offer instructions at qb64.org if you guys can offer me a set of them. When do these changes come into effect?

**Comment by uberhacker** [comment]
I can create pre-built binaries but you need this project installed on your Chromebook and simply install the package via `crew install qb64`.

**Comment by superloach** [comment]
Haha, I just saw you notice the PR got merged, I was about to tell you 😝

**Comment by uberhacker** [comment]
No worries at all.  I'm glad you are interested in this project.  Your feedback is greatly appreciated.

**Comment by superloach** [comment]
So how do you plan on presenting this on [qb64.org](https://qb64.org/)?
I just scrapped together this example, I feel like a link to that comment would be sufficient
```html
<td align="center" width="25%">
  <a href="https://github.com/skycocker/chromebrew/issues/2569#issuecomment-414528254" target="">
    <img width="65px" height="65px" src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a5/Google_Chrome_icon_%28September_2014%29.svg/1200px-Google_Chrome_icon_%28September_2014%29.svg.png">
  </a>
  <br>
  <a href="https://github.com/skycocker/chromebrew/issues/2569#issuecomment-414528254" style="text-decoration:none" target="">
    Chrome OS
  </a>
</td>
```
Obviously, if you chose to present it this way you'd use a pre-sized locally hosted icon, but overall I feel like this looks good.

![image](https://user-images.githubusercontent.com/30514424/44377466-df35e380-a4ca-11e8-9b21-ce791c880967.png)

(p.s. - sorry if I'm entering nitpick territory 😅)

**Comment by FellippeHeitor** [comment]
Hahaha, I was just creating a mockup now... this icon had come up in my searches: 

![](https://upload.wikimedia.org/wikipedia/commons/thumb/9/9d/Chromebook.svg/200px-Chromebook.svg.png)

Either way, because this is not a direct download, I thought of adding it as a smaller icon and an extra paragraph in that session with the summarized instructions. Sound good?

**Comment by superloach** [comment]
Sounds good! Good idea to use Chromium instead of Chrome, too.

**Comment by uberhacker** [comment]
@superloach: This is awesome what you are doing to promote Chromebrew.

**Comment by FellippeHeitor** [comment]
@superloach maybe your idea does look better to keep the general alignment there. Wanna give it a go?
https://github.com/FellippeHeitor/qb64.org/issues/1

**Comment by superloach** [comment]
Haha, no problem @uberhacker! Always glad to help share good projects, especially if I find them useful.

**Comment by uberhacker** [comment]
Fixed in #2577.
Fixed in #2578.
